src/_cffi_src/openssl/src/osrandom_engine.c

```diff
@@ -47,24 +47,44 @@ static int osrandom_rand_status(void) {
     }
 }
 #else
-static int urandom_fd = -1;
+static struct {
+    int fd;
+    dev_t st_dev;
+    ino_t st_ino;
+} urandom_cache = { -1 };
 
 static int osrandom_finish(ENGINE *e);
 
 static int osrandom_init(ENGINE *e) {
-    if (urandom_fd > -1) {
+    struct stat st;
+
+    if (urandom_cache.fd > -1) {
         return 1;
     }
-    urandom_fd = open("/dev/urandom", O_RDONLY);
-    if (urandom_fd > -1) {
-        int flags = fcntl(urandom_fd, F_GETFD);
+    urandom_cache.fd = open("/dev/urandom", O_RDONLY);
+    if (urandom_cache.fd > -1) {
+        int flags = fcntl(urandom_cache.fd, F_GETFD);
         if (flags == -1) {
             osrandom_finish(e);
             return 0;
-        } else if (fcntl(urandom_fd, F_SETFD, flags | FD_CLOEXEC) == -1) {
+        } else if (fcntl(urandom_cache.fd, F_SETFD, flags | FD_CLOEXEC) == -1) {
             osrandom_finish(e);
             return 0;
         }
+        errno = 0;
+        if (fstat(urandom_cache.fd, &st) == -1) {
+            /* As long as fstat failed for a reason that's not a bad file
+             * descriptor, call osrandom_finish to close the fd */
+            if (errno != EBADF) {
+                osrandom_finish(e);
```

Comment:
I think this is my fault for asking if there were any error cases where we wouldn't want to call `osrandom_finish`.